File: src/harness/llm.py
from __future__ import annotations

import json
import os
from getpass import getpass
from pathlib import Path
from typing import Any, Dict, List
from urllib.error import HTTPError, URLError
from urllib.request import Request, urlopen


class BaseLLM:
    """Abstract base class for LLM backends.

    Subclass this and implement :meth:`generate` to plug in any language model.
    """

    def generate(self, working_memory: Dict[str, Any]) -> Dict[str, Any]:
        """Generate the next agent action.

        Args:
            working_memory: Dict produced by :class:`~harness.memory.MemoryManager`
                containing ``goal``, ``context``, ``summary_memory``, ``history``,
                and optionally ``schema_feedback``.

        Returns:
            A dict that conforms to one of the two action schemas::

                {"type": "tool_call", "tool_name": "...", "arguments": {...}}
                {"type": "final_response", "content": "..."}

        Raises:
            NotImplementedError: Must be overridden by subclasses.
        """
        raise NotImplementedError


class ScriptedLLM(BaseLLM):
    """Deterministic LLM stub for testing the harness loop."""

    def __init__(self, script: List[Dict[str, Any]]) -> None:
        self._script = script
        self._index = 0

    def generate(self, working_memory: Dict[str, Any]) -> Dict[str, Any]:
        _ = working_memory
        if self._index >= len(self._script):
            return {
                "type": "final_response",
                "content": "Script exhausted. Stopping safely.",
            }
        action = self._script[self._index]
        self._index += 1
        return action


class RuleBasedLLM(BaseLLM):
    """Very small heuristic model for demo runs without external APIs."""

    def generate(self, working_memory: Dict[str, Any]) -> Dict[str, Any]:
        goal = str(working_memory.get("goal", "")).lower()
        history = working_memory.get("history", [])

        if "add" in goal or "sum" in goal:
            if not history:
                return {
                    "type": "tool_call",
                    "tool_name": "add",
                    "arguments": {"a": 2, "b": 3},
                }
            last_obs = str(history[-1].get("observation", ""))
            return {
                "type": "final_response",
                "content": f"Done. Computation result: {last_obs}",
            }

        if "time" in goal:
            if not history:
                return {
                    "type": "tool_call",
                    "tool_name": "utc_now",
                    "arguments": {},
                }
            return {
                "type": "final_response",
                "content": f"Current UTC time observed: {history[-1].get('observation')}",
            }

        return {
            "type": "final_response",
            "content": "No tool needed. Here is the direct answer.",
        }


class DeepSeekLLM(BaseLLM):
    def __init__(
        self,
        *,
        api_key: str | None = None,
        model: str = "deepseek-chat",
        base_url: str = "https://api.deepseek.com",
        env_path: str | Path | None = None,
        transport: Any | None = None,
    ) -> None:
        self.api_key = api_key
        self.model = model
        self.base_url = base_url.rstrip("/")
        self.env_path = Path(env_path) if env_path is not None else self._default_env_path()
        self.transport = transport or self._default_transport

    def generate(self, working_memory: Dict[str, Any]) -> Dict[str, Any]:
        api_key = self._resolve_api_key()
        payload = {
            "model": self.model,
            "messages": self._build_messages(working_memory),
            "temperature": 0.1,
        }
        response = self.transport(payload, api_key, self.base_url)
        return self._parse_response(response)

    def _resolve_api_key(self) -> str:
        api_key = self.api_key or os.environ.get("DEEPSEEK_API_KEY", "").strip()
        if not api_key:
            api_key = self._read_env_file_value("DEEPSEEK_API_KEY")
        if api_key:
            self.api_key = api_key
            return api_key

        prompted = getpass("DeepSeek API key not found. Enter DEEPSEEK_API_KEY: ").strip()
        if not prompted:
            raise ValueError("DeepSeek API key is required to run this entrypoint.")
        self._write_env_file_value("DEEPSEEK_API_KEY", prompted)
        os.environ["DEEPSEEK_API_KEY"] = prompted
        self.api_key = prompted
        return prompted

    @staticmethod
    def _default_env_path() -> Path:
        return Path(__file__).resolve().parents[2] / ".env"

    def _read_env_file_value(self, key: str) -> str:
        if not self.env_path.exists():
            return ""
        for line in self.env_path.read_text(encoding="utf-8").splitlines():
            stripped = line.strip()
            if not stripped or stripped.startswith("#") or "=" not in stripped:
                continue
            name, value = stripped.split("=", 1)
            if name.strip() == key:
                return value.strip().strip('"').strip("'")
        return ""

    def _write_env_file_value(self, key: str, value: str) -> None:
        lines: List[str] = []
        updated = False
        if self.env_path.exists():
            lines = self.env_path.read_text(encoding="utf-8").splitlines()

        next_lines: List[str] = []
        for line in lines:
            stripped = line.strip()
            if stripped and not stripped.startswith("#") and "=" in stripped:
                name, _ = stripped.split("=", 1)
                if name.strip() == key:
                    next_lines.append(f"{key}={value}")
                    updated = True
                    continue
            next_lines.append(line)

        if not updated:
            next_lines.append(f"{key}={value}")

        self.env_path.parent.mkdir(parents=True, exist_ok=True)
        self.env_path.write_text("\n".join(next_lines).rstrip() + "\n", encoding="utf-8")

    def _build_messages(self, working_memory: Dict[str, Any]) -> List[Dict[str, str]]:
        system_prompt = (
            "You are the LLM engine inside a tool-using agent harness. "
            "Return JSON only. "
            "Use exactly one of these shapes: "
            '{"type":"tool_call","tool_name":"<tool>","arguments":{...}} '
            'or {"type":"final_response","content":"<answer>"}. '
            "Available tools: echo(text), add(a,b), utc_now(), "
            "write_text_file(path, content), "
            "read_file(path, offset?, limit?) for reading file contents, "
            "edit_file(path, old_text, new_text) for precise text replacement, "
            "bash(command, timeout?) for running shell commands."
        )
        user_prompt = json.dumps(working_memory, ensure_ascii=False, indent=2)
        return [
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": user_prompt},
        ]

    @staticmethod
    def _parse_response(response: Dict[str, Any]) -> Dict[str, Any]:
        choices = response.get("choices")
        if not isinstance(choices, list) or not choices:
            raise ValueError("DeepSeek response did not include any choices.")

        message = choices[0].get("message", {})
        content = message.get("content")
        if not isinstance(content, str) or not content.strip():
            raise ValueError("DeepSeek response content was empty.")

        try:
            parsed = json.loads(content)
        except json.JSONDecodeError:
            return {
                "type": "final_response",
                "content": content.strip(),
            }
        if not isinstance(parsed, dict):
            raise ValueError("DeepSeek response JSON must decode to an object.")
        return parsed

    @staticmethod
    def _default_transport(
        payload: Dict[str, Any],
        api_key: str,
        base_url: str,
    ) -> Dict[str, Any]:
        request = Request(
            url=f"{base_url}/chat/completions",
            data=json.dumps(payload).encode("utf-8"),
            headers={
                "Authorization": f"Bearer {api_key}",
                "Content-Type": "application/json",
            },
            method="POST",
        )
        try:
            with urlopen(request, timeout=60) as response:
                return json.loads(response.read().decode("utf-8"))
        except HTTPError as exc:
            body = exc.read().decode("utf-8", errors="replace")
            raise RuntimeError(f"DeepSeek API returned HTTP {exc.code}: {body}") from exc
        except URLError as exc:
            raise RuntimeError(f"DeepSeek API request failed: {exc.reason}") from exc

File: ts/src/llm.ts
/**
 * LLM base class and adapters.
 * Provides BaseLLM, ScriptedLLM, RuleBasedLLM, DeepSeekLLM, and buildSystemPrompt.
 */

import * as fs from "node:fs";
import * as path from "node:path";
import type { ToolSchema } from "./types.js";

// ---------------------------------------------------------------------------
// buildSystemPrompt
// ---------------------------------------------------------------------------

export function buildSystemPrompt(
  toolNames: string[],
  options?: {
    nativeToolUse?: boolean;
    extraSystemInstructions?: string;
  },
): string {
  const nativeToolUse = options?.nativeToolUse ?? false;
  const extraSystemInstructions = options?.extraSystemInstructions ?? "";

  const toolsDesc = toolNames.length > 0 ? toolNames.join(", ") : "(none)";

  let outputSection: string;
  if (nativeToolUse) {
    outputSection =
      "## Output Format\n" +
      "Use the provided tools to accomplish the task. When you have completed the " +
      "task or need to communicate with the user, respond with a text message.\n\n";
  } else {
    outputSection =
      "## Output Format\n" +
      "Return exactly one JSON object per turn. Use one of these two shapes:\n" +
      '- Tool call: {"type":"tool_call","tool_name":"<tool>","arguments":{...}}\n' +
      '- Final answer: {"type":"final_response","content":"<answer>"}\n' +
      "Do not wrap JSON in markdown fences or add any text outside the JSON object.\n\n";
  }

  let prompt =
    "You are a coding agent. Your goal is to autonomously complete programming tasks " +
    "given by the user. You operate inside a tool-using harness that validates your " +
    "output, executes tools, and feeds results back to you.\n\n" +
    outputSection +
    `## Available Tools\n${toolsDesc}\n\n` +
    "## Workflow Strategy\n" +
    "Follow this order when working on code:\n" +
    "1. **Discover** — Use grep_search or glob_files to locate relevant files and " +
    "symbols. Use list_directory to understand project layout.\n" +
    "2. **Understand** — Use read_file to examine the code you found. Read enough " +
    "context to be confident about the change.\n" +
    "3. **Modify** — Use edit_file for surgical changes to existing files. Use " +
    "write_file only to create new files. Prefer small, focused edits.\n" +
    "4. **Verify** — Use bash to run tests, linters, or type checkers to confirm " +
    "your change works. Always verify before declaring success.\n\n" +
    "## Minimal Change Principle\n" +
    "Only modify what is necessary to complete the task. Do not refactor unrelated " +
    "code, rename variables for style, or reorganize imports unless the task " +
    "explicitly requires it.\n\n" +
    "## Error Recovery\n" +
    "When a tool call fails:\n" +
    "- Read the error message carefully and diagnose the root cause.\n" +
    "- Do NOT retry the exact same call. Change your approach: try a different " +
    "search pattern, fix the path, adjust the arguments, or gather more context " +
    "first.\n" +
    "- If you are stuck after 2-3 attempts, explain what you tried and why it " +
    "failed in a final_response.\n\n" +
    "## Context Markers\n" +
    "When you discover important information, prefix it with a marker so it " +
    "survives memory compression:\n" +
    "- `constraint:` for constraints or invariants you must respect.\n" +
    "- `todo:` for pending work items.\n" +
    "- `evidence:` for key findings (e.g., root cause of a bug).\n\n" +
    "## Safety Boundaries\n" +
    "Never execute destructive shell commands such as `rm -rf /`, " +
    "`git push --force`, `git reset --hard`, or anything that deletes data " +
    "or force-pushes to a remote. If the task seems to require a dangerous " +
    "operation, ask the user for confirmation in a final_response instead.";

  if (toolNames.includes("save_memory") && toolNames.includes("recall_memory")) {
    prompt +=
      "\n\n## Memory\n" +
      "You have persistent memory across sessions. Use save_memory to store important " +
      "discoveries (project conventions, test commands, architecture decisions). Use " +
      "recall_memory to retrieve previously saved knowledge.";
  }

  if (toolNames.includes("spawn_agent")) {
    prompt +=
      "\n\n## Sub-Agents\n" +
      "Use spawn_agent to delegate sub-tasks to specialized child agents. " +
      "The child runs to completion and returns a summary. " +
      "Check the context for available agent definitions.";
  }

  if (toolNames.includes("use_skill")) {
    prompt +=
      "\n\n## Skills\n" +
      "Use use_skill to look up reusable prompt templates for common tasks. " +
      "Check the context for available skills.";
  }

  if (extraSystemInstructions) {
    prompt += `\n\n## Additional Instructions\n${extraSystemInstructions}`;
  }

  return prompt;
}

// ---------------------------------------------------------------------------
// BaseLLM
// ---------------------------------------------------------------------------

export abstract class BaseLLM {
  onToken?: (token: string) => void;

  abstract generate(workingMemory: Record<string, unknown>): Promise<Record<string, unknown>>;

  setToolSchemas?(schemas: ToolSchema[]): void;
}

// ---------------------------------------------------------------------------
// ScriptedLLM
// ---------------------------------------------------------------------------

export class ScriptedLLM extends BaseLLM {
  private _script: Record<string, unknown>[];
  private _index = 0;

  constructor(script: Record<string, unknown>[]) {
    super();
    this._script = script;
  }

  async generate(_workingMemory: Record<string, unknown>): Promise<Record<string, unknown>> {
    if (this._index >= this._script.length) {
      return {
        type: "final_response",
        content: "Script exhausted. Stopping safely.",
      };
    }
    const action = this._script[this._index];
    this._index += 1;
    return action;
  }
}

// ---------------------------------------------------------------------------
// RuleBasedLLM
// ---------------------------------------------------------------------------

export class RuleBasedLLM extends BaseLLM {
  async generate(workingMemory: Record<string, unknown>): Promise<Record<string, unknown>> {
    const goal = String(workingMemory.goal ?? "").toLowerCase();
    const history = (workingMemory.history as Record<string, unknown>[]) ?? [];

    if (goal.includes("add") || goal.includes("sum")) {
      if (history.length === 0) {
        return {
          type: "tool_call",
          tool_name: "add",
          arguments: { a: 2, b: 3 },
        };
      }
      const lastObs = String(history[history.length - 1].observation ?? "");
      return {
        type: "final_response",
        content: `Done. Computation result: ${lastObs}`,
      };
    }

    if (goal.includes("time")) {
      if (history.length === 0) {
        return {
          type: "tool_call",
          tool_name: "utc_now",
          arguments: {},
        };
      }
      return {
        type: "final_response",
        content: `Current UTC time observed: ${history[history.length - 1].observation}`,
      };
    }

    return {
      type: "final_response",
      content: "No tool needed. Here is the direct answer.",
    };
  }
}

// ---------------------------------------------------------------------------
// DeepSeekLLM
// ---------------------------------------------------------------------------

/** Transport function signature: (url, requestInit) => response-like object */
export type TransportFn = (
  url: string,
  options: RequestInit,
) => Promise<{ status: number; ok: boolean; json: () => Promise<unknown> }>;

export class DeepSeekLLM extends BaseLLM {
  apiKey: string | undefined;
  model: string;
  baseUrl: string;
  envPath: string;
  private _transport: TransportFn;
  private _toolNames: string[] = [];

  constructor(options?: {
    apiKey?: string;
    model?: string;
    baseUrl?: string;
    envPath?: string;
    transport?: TransportFn;
  }) {
    super();
    this.apiKey = options?.apiKey;
    this.model = options?.model ?? "deepseek-chat";
    this.baseUrl = (options?.baseUrl ?? "https://api.deepseek.com").replace(/\/+$/, "");
    this.envPath = options?.envPath ?? this._defaultEnvPath();
    this._transport = options?.transport ?? this._defaultTransport;
  }

  setToolSchemas(schemas: ToolSchema[]): void {
    this._toolNames = schemas.map((s) => s.name);
  }

  async generate(workingMemory: Record<string, unknown>): Promise<Record<string, unknown>> {
    const apiKey = this._resolveApiKey();
    const payload = {
      model: this.model,
      messages: this._buildMessages(workingMemory),
      temperature: 0.1,
    };

    const maxRetries = 2;
    for (let attempt = 0; attempt <= maxRetries; attempt++) {
      try {
        const response = await this._transport(`${this.baseUrl}/chat/completions`, {
          method: "POST",
          headers: {
            Authorization: `Bearer ${apiKey}`,
            "Content-Type": "application/json",
          },
          body: JSON.stringify(payload),
        });

        if (!response.ok) {
          const errorMsg = `DeepSeek API returned HTTP ${response.status}`;
          if (this._isRetryableError(errorMsg) && attempt < maxRetries) {
            const wait = 2 ** attempt * 1000;
            await this._sleep(wait);
            continue;
          }
          throw new Error(errorMsg);
        }

        const data = (await response.json()) as Record<string, unknown>;
        return this._parseResponse(data);
      } catch (err) {
        if (err instanceof Error) {
          const msg = err.message;
          if (this._isRetryableError(msg) && attempt < maxRetries) {
            const wait = 2 ** attempt * 1000;
            await this._sleep(wait);
            continue;
          }
        }
        throw err;
      }
    }

    // Should never reach here, but TypeScript needs it
    throw new Error("DeepSeek API: max retries exhausted");
  }

  private _sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }

  private _isRetryableError(errorMsg: string): boolean {
    const match = errorMsg.match(/HTTP (\d{3})/);
    if (match) {
      const code = parseInt(match[1], 10);
      return code === 429 || code >= 500;
    }
    if (errorMsg.toLowerCase().includes("request failed")) {
      return true;
    }
    return false;
  }

  private _resolveApiKey(): string {
    let apiKey = this.apiKey || process.env.DEEPSEEK_API_KEY?.trim() || "";
    if (!apiKey) {
      apiKey = this._readEnvFileValue("DEEPSEEK_API_KEY");
    }
    if (apiKey) {
      this.apiKey = apiKey;
      return apiKey;
    }

    // In the TS version, we throw instead of prompting interactively
    // (interactive prompt would require readline which complicates testing)
    throw new Error(
      "DeepSeek API key is required. Set DEEPSEEK_API_KEY environment variable or provide it in .env file.",
    );
  }

  private _defaultEnvPath(): string {
    // Project root .env — 3 levels up from src/llm.ts
    return path.resolve(__dirname, "..", "..", ".env");
  }

  private _readEnvFileValue(key: string): string {
    try {
      if (!fs.existsSync(this.envPath)) {
        return "";
      }
      const content = fs.readFileSync(this.envPath, "utf-8");
      for (const line of content.split("\n")) {
        const stripped = line.trim();
        if (!stripped || stripped.startsWith("#") || !stripped.includes("=")) {
          continue;
        }
        const eqIndex = stripped.indexOf("=");
        const name = stripped.slice(0, eqIndex).trim();
        if (name === key) {
          let value = stripped.slice(eqIndex + 1).trim();
          // Strip surrounding quotes
          if (
            (value.startsWith('"') && value.endsWith('"')) ||
            (value.startsWith("'") && value.endsWith("'"))
          ) {
            value = value.slice(1, -1);
          }
          return value;
        }
      }
    } catch {
      // File read error — treat as missing
    }
    return "";
  }

  private _buildMessages(
    workingMemory: Record<string, unknown>,
  ): Array<{ role: string; content: string }> {
    const toolNames =
      this._toolNames.length > 0
        ? this._toolNames
        : ["echo", "add", "utc_now", "write_text_file"];
    const systemPrompt = buildSystemPrompt(toolNames);
    const userPrompt = JSON.stringify(workingMemory, null, 2);
    return [
      { role: "system", content: systemPrompt },
      { role: "user", content: userPrompt },
    ];
  }

  private _parseResponse(response: Record<string, unknown>): Record<string, unknown> {
    const choices = response.choices;
    if (!Array.isArray(choices) || choices.length === 0) {
      throw new Error("DeepSeek response did not include any choices.");
    }

    const message = (choices[0] as Record<string, unknown>).message as
      | Record<string, unknown>
      | undefined;
    const content = message?.content;
    if (typeof content !== "string" || !content.trim()) {
      throw new Error("DeepSeek response content was empty.");
    }

    try {
      const parsed = JSON.parse(content);
      if (typeof parsed !== "object" || parsed === null || Array.isArray(parsed)) {
        throw new Error("DeepSeek response JSON must decode to an object.");
      }
      return parsed as Record<string, unknown>;
    } catch (err) {
      if (err instanceof SyntaxError) {
        return {
          type: "final_response",
          content: content.trim(),
        };
      }
      throw err;
    }
  }

  private async _defaultTransport(
    url: string,
    options: RequestInit,
  ): Promise<{ status: number; ok: boolean; json: () => Promise<unknown> }> {
    const response = await fetch(url, options);
    return {
      status: response.status,
      ok: response.ok,
      json: () => response.json(),
    };
  }
}
